### PR TITLE
Ensure Process.MainWindowTitle and Process.Responding is refreshed

### PR DESCRIPTION
**Pull request by JeroenOortwijn**
PR for issue #36768.

It should fix the issue, but I'm unable to run the tests because opening the solution leads to the following NuGet error:
```
Error occurred while restoring NuGet packages: Invalid restore input. Duplicate frameworks found: 'net5.0, net5.0, net5.0, net5.0, net5.0, net5.0'. Input files: E:\Repos\JeroenOortwijn\dotnet\runtime\src\libraries\System.Diagnostics.Process\src\System.Diagnostics.Process.csproj.
```
(Which, I think, is tracked in #34173.)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=38385) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=38385) before we can accept your contribution.<br/><br/>:x: JeroenOortwijn [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=38385)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=38385) it.</sub>

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 1]
```diff
@@ -1593,6 +1593,87 @@ public void MainWindowHandle_GetWithGui_ShouldRefresh_Windows()
...
+        public void MainWindowTitle_GetWithGui_ShouldRefresh_Windows()
+        {
+            const string ExePath = "notepad.exe";
+            Assert.True(IsProgramInstalled(ExePath));
+
+            using (Process process = Process.Start(ExePath))
+            {
+                try
+                {
+                    Assert.Equal(string.Empty, process.MainWindowTitle);
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1609]: Why do we expect MainWindowTitle to start as empty and then be non-empty after a Refresh?

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 2]
```diff
@@ -1593,6 +1593,87 @@ public void MainWindowHandle_GetWithGui_ShouldRefresh_Windows()
...
+                        process.Refresh();
+                        if (!process.Responding)
+                        {
+                            break;
+                        }
+
+                        Thread.Sleep(100);
+                    }
+
+                    Assert.False(process.Responding);
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1665]: Similar question; why do we expect Responding to be false after a Refresh?

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 3]
```diff
@@ -1593,6 +1593,87 @@ public void MainWindowHandle_GetWithGui_ShouldRefresh_Windows()
...
+        {
+            const string ExePath = @"C:\Program Files\Windows NT\Accessories\wordpad.exe";
+            Assert.True(IsProgramInstalled(ExePath));
+
+            string dummyFilePath = $@"{Path.GetTempPath()}dummy_file";
+
+            var dummyFile = new FileStream(dummyFilePath, System.IO.FileMode.Create);
+            _ = dummyFile.Seek(2048L * 1024 * 1024, SeekOrigin.Begin);
+            dummyFile.WriteByte(0);
+            dummyFile.Close();
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1646]: Woh, this is creating a 2GB temporary file?  Even in outerloop that's probably something to be avoided.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs (hunk 1 above)
Comment (JeroenOortwijn) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1609]: I see this behaviour with the test program that I've written:
https://github.com/JeroenOortwijn/MainWindowTitle-test

With .Net Framework 4.8, the first `Process.MainWindowTitle` call (immediately after Notepad is started) returns an empty string.
As soon as the window is visible, it returns the contents of the window title.

Due to issue #36768, `Process.MainWindowTitle` keeps returning an empty string on .Net Core 3.1 and .Net 5.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs (hunk 2 above)
Comment (JeroenOortwijn) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1665]: I see this behaviour with the test program that I've written:
https://github.com/JeroenOortwijn/Responding-test

With .Net Framework 4.8, the first `Process.Responding` call (immediately after Wordpad is started) returns true.
After a couple of seconds, Wordpad is so busy loading the document that the title bar shows "(Not Responding)". At that point, `Process.Responding` returns false.

Due to issue #36768, `Process.Responding` keeps returning true on .Net Core 3.1 and .Net 5.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs (hunk 3 above)
Comment (JeroenOortwijn) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1646]: You're right, the file is quite big.
But I want to make sure that even on the fastest machine, Wordpad goes into a "not responding" state.
If the file is too small, Wordpad will have loaded the file without getting into that state.

So, some kind of optimum file size has to be found.

**Comment by adamsitnik** [comment]
@stephentoub could you please respond to @JeroenOortwijn answers to your review comments?

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1609]: Ok. Thanks.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1646]: If we really need to test that, I suggest the test itself needs to be the thing that causes the non-responsiveness, e.g. the test would RemoteExecutor.Invoke code that would create a window that's not responsive.  We shouldn't depend on creating a big enough file that the particular implementation in wordpad causes the UI to hang.

Comment (eiriktsarpalis) [reply src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:1649]: @stephentoub you mean with something like windows forms? I tried the following
```csharp
static void RunNotRespondingApp()
{
    var form = new System.Windows.Forms.Form();
    form.Show();
    Thread.Sleep(60_000);
}

RemoteExecutor.Invoke(RunNotRespondingApp);
```
but for whatever reason it wasn't enough to simulate not responding state using RemoteExecutor. It does work without issue when I run as a standalone app.

**Comment by eiriktsarpalis** [comment]
I rebased and ran the tests locally. Can confirm that the changes fix the behaviour (matching it with .NET Framework behaviour). I would recommend merging this PR ahead of the 5.0 deadline, but we could perhaps disable the problematic tests (it's not run in CI anyway, and we generally have no test coverage for `Process.Responding`). cc @adamsitnik @danmosemsft

**Comment by danmosemsft** [comment]
>  I would recommend merging this PR ahead of the 5.0 deadline, but we could perhaps disable the problematic test

Seems reasonable to me.

**Review by eiriktsarpalis — APPROVED** [review]

**Comment by eiriktsarpalis** [comment]
Thanks @JeroenOortwijn for your contribution!

Fixes #36768.

**Comment by danmosemsft** [comment]
@dotnet/dnceng @jaredpar timeouts on “ Build OSX x64 release Runtime_Debug” and similar on a lot of.PRs right now having a significant impact. Can you help us understand why?

**Comment by jaredpar** [comment]
@danmosemsft 

Looking through the logs it seems like the problem is the C++ build is just taking about 70 minutes to complete on OSX. That is eating up most of the available time for the job.

**Comment by jakubstilec** [comment]
@danmosemsft  checking now

**Comment by jakubstilec** [comment]
Job timed out but it wasn't helix machine, but azdo macos machine.

**Comment by jakubstilec** [comment]
I don't see any active issue in azdo. Retried the job and checking.

**Comment by jakubstilec** [comment]
@danmosemsft now macox build job "Build product" has finished in ~56 minutes, previously it timed out in 2 hours. It seems that it was a temporary issue on AzDO side.

**Comment by jaredpar** [comment]
@eiriktsarpalis why are we merging changes when the build legs are timing out?

**Comment by eiriktsarpalis** [comment]
@jaredpar Apologies, I assumed by the conversation that you were discussing infra issues. Do you need me to revert?

**Comment by jaredpar** [comment]
@jakubstilec latest seems to indicate it's working again so lets keep for now and willmonitor

**Comment by MattGal** [comment]
> @danmosemsft
> 
> Looking through the logs it seems like the problem is the C++ build is just taking about 70 minutes to complete on OSX. That is eating up most of the available time for the job.

We've moving from MacStadium to GitHub-hosted machines; The new machines have fewer cores.  I believe the only option currently is to increase your timeouts but, as usual, I know who you can engage in conversation if you think you can talk them into making  beefier mac vms.

**Comment by danmosemsft** [comment]
@MattGal thanks. I think that's for @jaredpar to decide on based on monitoring he mentions..

**Comment by MattGal** [comment]
> @MattGal thanks. I think that's for @jaredpar to decide on based on monitoring he mentions..

Yes I did quote reply of him but it looked like I was talking at you, sorry.  The key takeaway is they know the new hardware is slower and they're not going back to the old hardware.

**Comment by danmosemsft** [comment]
@jaredpar is it simple to increase our timeouts? If so perhaps we should do that pre-emptively, given we expect CI to stay busy for the near future?

**Pull request merged**